### PR TITLE
dbcache performance enhancements

### PR DESCRIPTION
**Pull request by ptschip**
This is currently fully functional, I just need to update some of the unit tests and should have this finished in the morning sometime.

I'm seeing very good performance even with a small 100MB dbcache.  Previously the dbcache was just getting cleared entirely every time the max limit was exceeded causing the next few blocks to be quite slow at validating while the cache was rebuilding.  All we do here is, when flushing the chainstate, keep cache entries that are not marked DIRTY and we also trim non-dirty entries if we exceed the max.

If you want to test with it use the following to see the log entries before and after flushing the chainstate.

debug=coindb
debug=thin

### CODE REVIEW

**Comment by theslagter** [comment]
ptschip why you keep working on bitcoin unlimited? how many hours you spend on this per day?

**Comment by ptschip** [comment]
taking out of WIP, all tests are passing now.

**Comment by sickpig** [comment]
@ptschip need to be rebased cause it miss all the commit made from since Jun 16, namely:

```
e209fb8 gandrewstone    Merge pull request #675 from kyuupichan/CNetAddrService
046e241 gandrewstone    Merge pull request #674 from Justaphf/devWinDebug
be7c810 gandrewstone    Merge pull request #683 from kyuupichan/PV-fix
9fcf4b6 gandrewstone    Merge pull request #685 from ptschip/dev_headers
ba34c28 Peter Tschipper Remove DOS from unconnecting thinblock
4abd3b2 Peter Tschipper Do not continue if the header will not connect
fa9ab32 Neil Booth      PV: shut down threads only if initialized
79c23a0 Neil Booth      Move CNetAddr and CService to netaddress.{cpp,h}
b1d866b Justaphf        Fix debug build Pair ambiguity on Windows
```

**Comment by ptschip** [comment]
@sickpig ok rebased.

**Comment by sickpig** [comment]
@ptschip thanks

**Comment by ptschip** [comment]
@sickpig i added one small commit...it won't affect your general testing; it only impacts reindexing.

**Comment by ptschip** [comment]
rebased

File: src/main.cpp [hunk 1]
```diff
@@ -1221,15 +1221,6 @@ bool AcceptToMemoryPoolWorker(CTxMemPool &pool,
...
 
-            // do we already have it?
-            bool fHadTxInCache = pcoinsTip->HaveCoinsInCache(hash);
-            if (view.HaveCoins(hash))
-            {
-                if (!fHadTxInCache)
-                    vHashTxnToUncache.push_back(hash);
-                return state.Invalid(false, REJECT_ALREADY_KNOWN, "txn-already-known");
-            }
-
```
Comment (gandrewstone) [inline src/main.cpp:1232]: why is this deleted?

File: src/main.cpp [hunk 2]
```diff
@@ -3219,16 +3207,19 @@ bool static ConnectTip(CValidationState &state,
...
     int64_t nTime4 = GetTimeMicros();
     nTimeFlush += nTime4 - nTime3;
     LogPrint("bench", "  - Flush: %.2fms [%.2fs]\n", (nTime4 - nTime3) * 0.001, nTimeFlush * 0.000001);
-    // Write the chain state to disk, if necessary.
-    if (!FlushStateToDisk(state, FLUSH_STATE_IF_NEEDED))
-        return false;
+    // Write the chain state to disk, if necessary, and only during IBD, reindex, or importing.
+    if (!IsChainNearlySyncd() || fReindex || fImporting)
+        if (!FlushStateToDisk(state, FLUSH_STATE_IF_NEEDED))
+            return false;
```
Comment (gandrewstone) [inline src/main.cpp:3217]: why is this not necessary during normal operation?

File: src/main.cpp (hunk 1 above)
Comment (ptschip) [reply src/main.cpp:1232]: I took this out because it makes no sense...why would we delete a coin from the dbcache just because we already have it there...now we have it, now we don't...furthermore, why would we reject a txn based on what the state of the dbacahe is...the coin could be there already from finishing up IBD and there are ouputs left over in the dbcache or perhaps in the future for some other reason...if we're not going to allow a txn in the mempool it has to be because it's either an invalid txn or because we already have it in the mempool...checking the dbcache is not a good or reliable test IMO for whether the txn is in the mempool.

File: src/main.cpp (hunk 2 above)
Comment (ptschip) [reply src/main.cpp:3217]: There are two reasons:  1) We flush ALWAYS now after every new block is processed (when processnewblock returns, at the end of HandleBlockMessageThread) during normal operation so checking again is redundant, furthermore, it injects performance timing issues if we flush at that point during ConnectTip  it can add a significant amount of time to the processing of a block in our statistics, when acutally, it has nothing to do with block processsing and should be excluded from the timing.

File: src/main.cpp (hunk 1 above)
Comment (gandrewstone) [reply src/main.cpp:1232]: I think that this is an optimization.  it basically says "if the tx is already in the utxo cache, then clearly we already know about it".  IDK why you'd subsequently want to uncache the tx though... anyway I am fine with leaving it removed
